File: frontend/src/app/core/services/visits/visits.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';
import { API_URL } from '../../../utils/constants';
import {
  ScheduleResponse,
  ScheduleVisitRequest,
} from '../../models/schedule.model';
import {
  SingleVisitResponse,
  UpcomingVisitsResponse,
  VisitResponse,
} from '../../models/visit.model';
import { FinishVisitResponse } from './../../models/visit.model';

@Injectable({
  providedIn: 'root',
})
export class VisitsService {
  private http = inject(HttpClient);

  public getUpcomingVisits() {
    return this.http
      .get<UpcomingVisitsResponse>(`${API_URL}/users/me/visits?upcoming=true`, {
        withCredentials: true,
      })
      .pipe(map((response) => response.visits ?? []));
  }

  public getAllVisits() {
    return this.http
      .get<UpcomingVisitsResponse>(`${API_URL}/users/me/visits`, {
        withCredentials: true,
      })
      .pipe(
        map((response) => {
          return response.visits ?? [];
        }),
      );
  }

  public getVisitDetails(visitId: string) {
    return this.http.get<SingleVisitResponse>(`${API_URL}/visits/${visitId}`, {
      withCredentials: true,
    });
  }

  public cancelVisit(visitId: string) {
    return this.http.delete(`${API_URL}/visits/${visitId}`, {
      withCredentials: true,
    });
  }

  public scheduleVisit(scheduleVisit: ScheduleVisitRequest) {
    return this.http.post(`${API_URL}/visits/add`, scheduleVisit, {
      withCredentials: true,
    });
  }

  public rescheduleVisit(scheduleVisit: ScheduleVisitRequest, visitId: string) {
    return this.http.put(
      `${API_URL}/visits/${visitId}/reschedule`,
      {
        patientRemarks: scheduleVisit.patientRemarks,
        newschedule: scheduleVisit.scheduleID,
      },
      { withCredentials: true },
    );
  }

  public getDoctorSchedule(doctorId: string) {
    return this.http.get<ScheduleResponse>(
      `${API_URL}/doctors/${doctorId}/schedules`,
      {
        withCredentials: true,
      },
    );
  }

  public getDoctorVisitByDate(date: string): Observable<VisitResponse[]> {
    return this.http
      .get<UpcomingVisitsResponse>(
        `${API_URL}/doctors/me/visitsbydate/${date}`,
        {
          withCredentials: true,
        },
      )
      .pipe(map((response) => response.visits ?? []));
  }

  public finishVisit(
    finishVisit: FinishVisitResponse,
    visitId: string,
  ): Observable<void> {
    return this.http.put<void>(
      `${API_URL}/visits/${visitId}/complete`,
      {
        prescriptions: finishVisit.prescriptions,
        referrals: finishVisit.referrals,
        note: finishVisit.note,
      },
      { withCredentials: true },
    );
  }
}

File: mobile/app/src/main/java/com/medipath/core/services/VisitsService.kt
package com.medipath.core.services

import com.medipath.core.models.BookingRequest
import com.medipath.core.models.CompleteVisitRequest
import com.medipath.core.models.VisitDetailsResponse
import com.medipath.core.models.VisitsResponse
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface VisitsService {
    @GET("/api/users/me/visits")
    suspend fun getAllVisits(): Response<VisitsResponse>

    @GET("/api/users/me/visits")
    suspend fun getUpcomingVisits(@Query("upcoming") upcoming: String = "true"): Response<VisitsResponse>

    @GET("/api/visits/{id}")
    suspend fun getVisitDetails(@Path("id") visitId: String): Response<VisitDetailsResponse>

    @DELETE("/api/visits/{visitid}")
    suspend fun cancelVisit(@Path("visitid") visitId: String): Response<Unit>
    
    @POST("/api/visits/add")
    suspend fun bookAppointment(@Body bookingRequest: BookingRequest): Response<Unit>
    
    @PUT("/api/visits/{visitid}/reschedule")
    suspend fun rescheduleVisit(@Path("visitid") visitId: String, @Query("newschedule") newScheduleId: String): Response<Unit>
    
    @PUT("/api/visits/{visitid}/complete")
    suspend fun completeVisit(@Path("visitid") visitId: String, @Body request: CompleteVisitRequest): Response<Unit>
}